File: app/src/main/java/at/priv/graf/fragments/SectionListAdapter.java
package at.priv.graf.fragments;

import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import at.priv.graf.zazentimer.R;
import at.priv.graf.zazentimer.bo.Section;
import java.util.ArrayList;
import java.util.List;

public class SectionListAdapter extends RecyclerView.Adapter<SectionListAdapter.ViewHolder> {

    private List<Section> items = new ArrayList<>();
    private final OnItemClickListener clickListener;
    private final OnSectionActionListener actionListener;

    public interface OnItemClickListener {
        void onItemClick(Section section);
    }

    public interface OnSectionActionListener {
        void onDeleteSection(int position);
        void onDuplicateSection(int position);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView spinnerText1;
        final TextView spinnerText2;
        final ImageView dragHandle;
        final ImageButton sectionOverflow;

        ViewHolder(View view) {
            super(view);
            dragHandle = view.findViewById(R.id.dragHandle);
            spinnerText1 = view.findViewById(R.id.spinnerText1);
            spinnerText2 = view.findViewById(R.id.spinnerText2);
            sectionOverflow = view.findViewById(R.id.sectionOverflow);
        }
    }

    public SectionListAdapter(OnItemClickListener clickListener) {
        this(clickListener, null);
    }

    public SectionListAdapter(OnItemClickListener clickListener, OnSectionActionListener actionListener) {
        this.clickListener = clickListener;
        this.actionListener = actionListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.session_list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Section section = items.get(position);

        String name;
        if (section.name != null && section.name.length() > 0) {
            name = section.name;
        } else {
            name = holder.itemView.getContext().getString(R.string.unnamed);
        }

        String durationStr = section.getDurationString() + ", ";
        String info;
        if (section.bellcount == 1) {
            info = durationStr + holder.itemView.getContext().getString(R.string.section_info_string_1_sg);
        } else {
            String partial = durationStr + String.format(holder.itemView.getContext().getString(R.string.section_info_string_1_pl), section.bellcount) + " ";
            if (section.bellpause == 1) {
                info = partial + holder.itemView.getContext().getString(R.string.section_info_string_2_sg);
            } else {
                info = partial + String.format(holder.itemView.getContext().getString(R.string.section_info_string_2_pl), section.bellpause);
            }
        }

        holder.spinnerText1.setText(name);
        holder.spinnerText2.setText(info);

        holder.itemView.setOnClickListener(v -> {
            if (clickListener != null) {
                clickListener.onItemClick(items.get(holder.getBindingAdapterPosition()));
            }
        });

        holder.sectionOverflow.setOnClickListener(v -> {
            PopupMenu popup = new PopupMenu(v.getContext(), v);
            popup.getMenuInflater().inflate(R.menu.menu_section_card_actions, popup.getMenu());
            popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                @Override
                public boolean onMenuItemClick(MenuItem menuItem) {
                    int pos = holder.getBindingAdapterPosition();
                    if (pos == RecyclerView.NO_POSITION || actionListener == null) {
                        return false;
                    }
                    int id = menuItem.getItemId();
                    if (id == R.id.card_action_delete_section) {
                        actionListener.onDeleteSection(pos);
                        return true;
                    } else if (id == R.id.card_action_duplicate_section) {
                        actionListener.onDuplicateSection(pos);
                        return true;
                    }
                    return false;
                }
            });
            popup.show();
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void setItems(List<Section> newItems) {
        items = new ArrayList<>(newItems);
        notifyDataSetChanged();
    }

    public Section getItem(int position) {
        return items.get(position);
    }

    public Section removeItem(int position) {
        Section removed = items.remove(position);
        notifyItemRemoved(position);
        return removed;
    }

    public void insertItem(int position, Section section) {
        items.add(position, section);
        notifyItemInserted(position);
    }

    public List<Section> getItems() {
        return new ArrayList<>(items);
    }

    public void moveItem(int fromPosition, int toPosition) {
        Section item = items.remove(fromPosition);
        items.add(toPosition, item);
        notifyItemMoved(fromPosition, toPosition);
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/fragments/SectionListAdapter.kt
package at.priv.graf.zazentimer.fragments

import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView
import at.priv.graf.zazentimer.R
import at.priv.graf.zazentimer.bo.Section

class SectionListAdapter(
    private val clickListener: OnItemClickListener?,
    private val actionListener: OnSectionActionListener? = null,
) : RecyclerView.Adapter<SectionListAdapter.ViewHolder>() {
    interface OnItemClickListener {
        fun onItemClick(section: Section)
    }

    interface OnSectionActionListener {
        fun onDeleteSection(position: Int)

        fun onDuplicateSection(position: Int)
    }

    private var items: MutableList<Section> = ArrayList()

    class ViewHolder(
        view: View,
    ) : RecyclerView.ViewHolder(view) {
        val spinnerText1: TextView = view.findViewById(R.id.spinnerText1)
        val spinnerText2: TextView = view.findViewById(R.id.spinnerText2)
        val dragHandle: ImageView = view.findViewById(R.id.dragHandle)
        val sectionOverflow: ImageButton = view.findViewById(R.id.sectionOverflow)
    }

    @NonNull
    override fun onCreateViewHolder(
        @NonNull parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.session_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(
        @NonNull holder: ViewHolder,
        position: Int,
    ) {
        val section = items[position]

        val name: String =
            section.name?.takeIf { it.isNotEmpty() } ?: holder.itemView.context.getString(R.string.unnamed)

        val durationStr = "${section.getDurationString()}, "
        val info: String =
            if (section.bellcount == 1) {
                "$durationStr${holder.itemView.context.getString(R.string.section_info_string_1_sg)}"
            } else {
                val partial =
                    "$durationStr${String.format(holder.itemView.context.getString(R.string.section_info_string_1_pl), section.bellcount)} "
                if (section.bellpause == 1) {
                    "$partial${holder.itemView.context.getString(R.string.section_info_string_2_sg)}"
                } else {
                    "$partial${String.format(holder.itemView.context.getString(R.string.section_info_string_2_pl), section.bellpause)}"
                }
            }

        holder.spinnerText1.text = name
        holder.spinnerText2.text = info

        holder.itemView.setOnClickListener {
            clickListener?.onItemClick(items[holder.bindingAdapterPosition])
        }

        holder.sectionOverflow.setOnClickListener {
            val popup = PopupMenu(it.context, it)
            popup.menuInflater.inflate(R.menu.menu_section_card_actions, popup.menu)
            popup.setOnMenuItemClickListener(
                object : PopupMenu.OnMenuItemClickListener {
                    override fun onMenuItemClick(menuItem: MenuItem): Boolean {
                        val pos = holder.bindingAdapterPosition
                        if (pos == RecyclerView.NO_POSITION || actionListener == null) {
                            return false
                        }
                        val id = menuItem.itemId
                        if (id == R.id.card_action_delete_section) {
                            actionListener.onDeleteSection(pos)
                            return true
                        } else if (id == R.id.card_action_duplicate_section) {
                            actionListener.onDuplicateSection(pos)
                            return true
                        }
                        return false
                    }
                },
            )
            popup.show()
        }
    }

    override fun getItemCount(): Int = items.size

    fun setItems(newItems: List<Section>) {
        items = ArrayList(newItems)
        notifyDataSetChanged()
    }

    fun getItem(position: Int): Section = items[position]

    fun removeItem(position: Int): Section {
        val removed = items.removeAt(position)
        notifyItemRemoved(position)
        return removed
    }

    fun insertItem(
        position: Int,
        section: Section,
    ) {
        items.add(position, section)
        notifyItemInserted(position)
    }

    fun getItems(): List<Section> = ArrayList(items)

    fun moveItem(
        fromPosition: Int,
        toPosition: Int,
    ) {
        val item = items.removeAt(fromPosition)
        items.add(toPosition, item)
        notifyItemMoved(fromPosition, toPosition)
    }
}
